File: leetcode/GroupAnagrams.py
import os
import sys

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        sorts = {}
        for i, word in enumerate(strs):
            sorted_word = str(sorted(word))
            arr = sorts.get(sorted_word)
            if arr is None:
                arr = []
            arr.append(i)
            sorts.update({sorted_word: arr})
        
        res = []
        for value in sorts.values():
            sub_res = []
            for index in value:
                sub_res.append(strs[index])
            res.append(sub_res)
        print(res)
        

def rinput():
	data = input()
	return data.split(" ");

if __name__ == "__main__":
	solution = Solution()
	res = solution.groupAnagrams(["eat","tea","tan","ate","nat","bat"])
	print(res)

# input: 2,7,11,15 9

File: leetcode/leetbean/leetbean/src/main/java/org/strixsc/ArraysAndHashing/GroupAnagrams.java
package org.strixsc.ArraysAndHashing;

import java.util.*;
import java.util.stream.Collectors;

public class GroupAnagrams {
    public static class Solution {

        public static boolean isAnagram(String s, String t) {
            if (s.length() != t.length())
            {
                return false;
            }

            char[] S = s.toCharArray();
            Arrays.sort(S);
            char[] T = t.toCharArray();
            Arrays.sort(T);

            for (int i = 0; i < S.length; i++) {
                if (S[i] != T[i]) {
                    return false;
                }
            }

            return true;
        }

        public static List<List<String>> groupAnagrams(String[] strs) {
            Map<String, List<String>> hm = new HashMap<String, List<String>>();

            for (String string : strs)
            {
                char[] sortedChars = string.toCharArray();
                Arrays.sort(sortedChars);
                String sorted = new String(sortedChars);
                if (hm.containsKey(sorted)) {
                    hm.get(sorted).add(string);
                } else {
                    hm.put(new String(sorted), new ArrayList<>(Arrays.asList(string)));
                }
            }

            return new ArrayList<>(hm.values()).stream().filter(v -> !v.isEmpty()).collect(Collectors.toList());
        }
    }

    public static void main(String[] args) {
        System.out.println(Solution.groupAnagrams(new String[]{"eat","tea","tan","ate","nat","bat"}));
    }
}
